### PR TITLE
engine/message: notification queue refactor

### PR DESCRIPTION
**Pull request by mastercactapus**
- [x] Identified the issue which this PR solves.
- [x] Read the [**CONTRIBUTING**](https://github.com/target/goalert/blob/master/CONTRIBUTING.md) document.
- [x] Code builds clean without any errors or warnings.
- [x] Added appropriate tests for any new functionality.
- [x] All new and existing tests passed.
- [x] Added comments in the code, where necessary.
- [x] Ran `make check` to catch common errors. Fixed any that came up.

**Description:**
This PR updates the message module to handle queue prioritization entirely within the application.

This allows us to efficiently sort outgoing messages recursively to ensure the right message goes out when the outgoing rate is at-capacity.

- Rate limit for slack is now 5 per second
- Messages are now sent concurrently (up to 5 per type at a time)
- A single contact method can only receive 1 message per minute
- prioritization of "First Alert" per-service per 10-minute window
- prioritization is now unit-testable
- engine sendmessage greatly simplified
- added index on outgoing messages for (last_status, sent_at)

### CODE REVIEW

**Comment by mastercactapus** [comment]
The destination type isn't used with regard to prioritization, so we left it out of the `_Sort` test. The complicated thing that warranted a unit test was only the priority order so we just kept it simple.

It might be worth changing one of the sent messages to a different type though just to point-out/prove that all destinations are considered regardless of the value passed to `NextByType`. That's pretty easy to validate visually, but if anything having an outlier and comment would probably be useful documentation/context for someone looking at it in the future.

--- 

As far as the implementation, we used the built-in interfaces for sorting and stuck with bare slices because it was super easy to iterate on the behavior. I'm sure there are several optimizations we can make and better ways to implement it. 

Worth noting though: the most important thing is that the code is easy to read/audit and iterate on; performance itself isn't really a concern (well, within reason). It will also be changing a lot in the near future, so there's that too.

I did, however, run a quick benchmark locally. 30k messages in 150ms and 88 allocations (the max we've seen is 1.5k) -- so even as-is it's not doing too bad. 😂

File: engine/message/queue.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+		pjTypePriority := typePriority[pj.Type]
+		if pj.Type == TypeAlertNotification && q.serviceSent[pj.ServiceID].IsZero() {
+			pjTypePriority = 0
+		}
+
+		if piTypePriority != pjTypePriority {
+			return piTypePriority < pjTypePriority
+		}
+
+		if isLess, ok := q.userPriority(pi.UserID, pj.UserID); ok {
```
Comment (Auchindoun) [inline engine/message/queue.go]: Should we return out with an error if not ok, or resort to random ordering?

File: engine/message/db.go [hunk 2]
```diff
@@ -264,82 +265,89 @@ func NewDB(ctx context.Context, db *sql.DB, c *Config) (*DB, error) {
...
 	}
 	defer rows.Close()
 
-	var result []row
-	var r row
+	var result []Message
 	for rows.Next() {
-		var cmID, cmType, chID, chType sql.NullString
+		var msg Message
+		var cID, cValue, verifyID, userID, serviceID, cmType, chanType sql.NullString
```
[the remark this answers is not in the document]
Comment (mastercactapus) [reply engine/message/db.go]: It refers to either channel ID _or_ contact method ID based on the type (which isn't known until after scanning). Same for `cValue` it can be the channel value or the contact method value depending on type.

Only `cmType` and `chanType` are distinct.

[the remark this answers is not in the document]
Comment (Auchindoun) [reply engine/message/db.go]: Ah, is there some way we can denote that? Variable names in Go can be hard to follow, especially while reviewing/revisiting.

File: engine/message/queue.go (hunk 1 above)
Comment (mastercactapus) [reply engine/message/queue.go]: OK means "yes we have determined priority" but not an error scenario. Basically `ok` means `isLess` is valid. If `ok` is falls, then `isLess` should be ignored.

It's following the same pattern as type assertions and map existence checks:
```go
str, ok := interfaceVal.(string)

val, ok := someMap["value"]
```

File: engine/message/db.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mastercactapus) [reply engine/message/db.go]: Hmm... I think `destID` and `destValue` should work -- They are only used once a few lines down as `msg.Dest.ID` and `msg.Dest.Value`, respectively.
